src/lsm/zig_zag_merge.zig

```diff
@@ -0,0 +1,664 @@
+    /// before calling `peek()` again.
+    /// Returns `Empty` if the stream was fully consumed and reached the end.
+    comptime stream_peek: fn (
+        context: *Context,
+        stream_index: u32,
+    ) error{ Empty, Drained }!Key,
+    /// Consumes the current value and moves the stream identified by `stream_index`.
+    /// Pop is always called after `peek()`, it is not expected that the stream be `Empty`
+    /// or `Drained`.
+    comptime stream_pop: fn (context: *Context, stream_index: u32) Value,
+    /// Probes the stream identified by `stream_index` causing it to move to the next value such
+    /// as `>=key` or `<=key` depending on the iterator direction.
+    /// Should not be called when the current key already matches the probe.
+    /// The stream may become `Empty` or `Drained` _after_ probing.
+    comptime stream_probe: fn (context: *Context, stream_index: u32, key: Key) void,
+) type {
+    return struct {
+        const ZigZagMergeIterator = @This();
+        const BitSet = std.bit_set.IntegerBitSet(streams_max);
+
+        context: *Context,
+        streams_count: u32,
+        direction: Direction,
+        previous_key_popped: ?Key = null,
+
+        /// At least two scans are required for zig-zag merge.
+        pub fn init(
+            context: *Context,
+            streams_count: u32,
+            direction: Direction,
+        ) ZigZagMergeIterator {
+            assert(streams_count <= streams_max);
+            assert(streams_count > 1);
+
+            return .{
+                .context = context,
+                .streams_count = streams_count,
+                .direction = direction,
+            };
+        }
+
+        // Resets the iterator when the underlying streams are moved.
+        // It's not necessary for ZigZagMerge, but it follows the same API for all MergeIterators.
+        pub fn reset(it: *ZigZagMergeIterator) void {
+            _ = it;
+        }
+
+        pub fn pop(it: *ZigZagMergeIterator) error{Drained}!?Value {
+            while (try it.peek_key()) |key| {
+                const value = stream_pop(it.context, 0);
+                assert(key_from_value(&value) == key);
+                for (1..it.streams_count) |stream_index| {
+                    const value_other = stream_pop(it.context, @intCast(stream_index));
+                    assert(key_from_value(&value_other) == key);
+
+                    if (constants.verify) {
+                        // It's assumed that streams will produce the same value.
+                        assert(stdx.equal_bytes(Value, &value, &value_other));
+                    }
+                }
+
+                if (it.previous_key_popped) |previous| {
+                    switch (std.math.order(previous, key)) {
+                        .lt => assert(it.direction == .ascending),
+                        // Discard duplicate values.
+                        .eq => continue,
```

Comment:
Well, I can't foresee any concrete case, even something like `(A OR B) AND (A OR C)` would not produce a duplicated value.

However, since our binary search supports duplicated values, it would be reasonable to assume that `{1,1} ∩ {1,1} = {1}`. But I agree it's better to `unreachable` to avoid masking unintended behavior.